closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -343,40 +343,47 @@ int encrypt(byte[] in, int inOfs, int len) {
      */
     int encryptFinal(byte[] in, int inOfs, int len, byte[] out, int outOfs)
         throws IllegalBlockSizeException, ShortBufferException {
+        byte[] keyCopy;
+        byte[] ivCopy;
+        byte[] aad;
+        synchronized (this) {
+            if (len > MAX_BUF_SIZE - tagLenBytes) {
+                throw new ShortBufferException
+                    ("Can't fit both data and tag into one buffer");
+            }
 
-        if (len > MAX_BUF_SIZE - tagLenBytes) {
-            throw new ShortBufferException
-                ("Can't fit both data and tag into one buffer");
-        }
+            if (out.length - outOfs < (len + tagLenBytes)) {
+                throw new ShortBufferException("Output buffer too small");
+            }
 
-        if (out.length - outOfs < (len + tagLenBytes)) {
-            throw new ShortBufferException("Output buffer too small");
-        }
+            // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+            keyCopy = key.clone();
+            ivCopy = iv.clone();
```

Comment:
This should be delayed to avoid this work only to throw an exception.

Reply:
Okay, I moved them till the end of the synchronized block.